src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -3125,8 +3125,23 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
                                       ASR::down_cast<ASR::List_t>(type)->m_type, nullptr);
                 return false;
             } else if (ASR::is_a<ASR::Tuple_t>(*type)) {
+                int i = 0;
                 index = ASRUtils::EXPR(tmp);
-                int i = ASR::down_cast<ASR::IntegerConstant_t>(ASRUtils::EXPR(tmp))->m_n;
+                ASR::expr_t* val = ASRUtils::expr_value(index);
+                if (val && ASR::is_a<ASR::IntegerConstant_t>(*val)) {
```

Comment:
If `val` is `nullptr` then it should throw an error I guess. Runtime indexing isn't possible with tuples because then the type of `TupleItem` cannot be figured out at compile time so generating the node itself won't be possible.

Reply:
Yeah as llvm backend does not yet support variable index for tuples.... for eg these programs
```
def main0():
    t: tuple[i32, i32, i32] = (1, 2, 3)
    x: i32 = -1
    print(t[x])

main0()
```
So for these cases , we can see that `val` would turn out to be `nullptr` . Hence we could surely raise something here like `Runtime Indexing with Tuples is not yet supported` .

Reply:
After the latest commit we would get the following error raised for the program .
```
(lf) anutosh491@spbhat68:~/lpython/lpython$ ./src/bin/lpython examples/expr2.py 
semantic error: Runtime Indexing with Tuples is not yet supported
 --> examples/expr2.py:4:11
  |
4 |     print(t[x])
  |           ^^^^ 
```

Reply:
IMO, runtimes indexing with tuples can never be supported in a strictly typed compiler like LPython. For example C++ (a statically typed language) also doesn't support runtime indices with tuples. The reason is fairly simple if the index is runtime then type of the indexed item becomes runtime, which implies the destination variable where the indexed item will be stored has a runtime type but that's contradictory because the compiler is strictly typed i.e., you should know the type of each variable while writing the code.

**What should we do?** - An error message like, `"Runtime Indexing with " + ASRUtils::type_to_str_python(whatever_the_type_of_tuple_is) + "is not possible"`. That way user will know that it isn't possible to pass runtime indices to tuples like, `tuple[i32, str, ...]`.

Supporting `tuple[any, any, any]` will be a separate case which I think is different from what we are dealing with here. 

**Code**

```cpp
#include <tuple>
#include <iostream>
#include <string>
#include <stdexcept>

std::tuple<double, char, std::string> get_student(int id)
{
    switch (id)
    {
        case 0: return {3.8, 'A', "Lisa Simpson"};
        case 1: return {2.9, 'C', "Milhouse Van Houten"};
        case 2: return {1.7, 'D', "Ralph Wiggum"};
        case 3: return {0.6, 'F', "Bart Simpson"};
    }

    throw std::invalid_argument("id");
}

int main()
{

    int i, j, k;
    const auto student0 = get_student(0);
    std::cin >> i >> j >> k;
    std::cout << std::get<i>(student0)
              << ", " << std::get<j>(student0)
              << ", " << std::get<k>(student0) << '\n';
}
```

**Shell Output**

```zsh
(lp) 20:07:50:~/lpython_project/lpython % clang++ /Users/czgdp1807/lpython_project/debug.cpp -"std=c++17"
/Users/czgdp1807/lpython_project/debug.cpp:25:18: error: no matching function for call to 'get'
    std::cout << std::get<i>(student0)
                 ^~~~~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__tuple:226:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(array<_Tp, _Size>&) _NOEXCEPT;
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__tuple:231:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(const array<_Tp, _Size>&) _NOEXCEPT;
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__tuple:237:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(array<_Tp, _Size>&&) _NOEXCEPT;
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__tuple:242:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(const array<_Tp, _Size>&&) _NOEXCEPT;
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:491:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(pair<_T1, _T2>& __p) _NOEXCEPT
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:499:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(const pair<_T1, _T2>& __p) _NOEXCEPT
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:508:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(pair<_T1, _T2>&& __p) _NOEXCEPT
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:516:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(const pair<_T1, _T2>&& __p) _NOEXCEPT
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:525:17: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1 & get(pair<_T1, _T2>& __p) _NOEXCEPT
                ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:532:23: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1 const & get(pair<_T1, _T2> const& __p) _NOEXCEPT
                      ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:539:18: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1 && get(pair<_T1, _T2>&& __p) _NOEXCEPT
                 ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:546:24: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1 const && get(pair<_T1, _T2> const&& __p) _NOEXCEPT
                       ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:553:17: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1 & get(pair<_T2, _T1>& __p) _NOEXCEPT
                ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:560:23: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1 const & get(pair<_T2, _T1> const& __p) _NOEXCEPT
                      ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:567:18: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1 && get(pair<_T2, _T1>&& __p) _NOEXCEPT
                 ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__utility/pair.h:574:24: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1 const && get(pair<_T2, _T1> const&& __p) _NOEXCEPT
                       ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/tuple:1129:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(tuple<_Tp...>& __t) _NOEXCEPT
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/tuple:1138:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(const tuple<_Tp...>& __t) _NOEXCEPT
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/tuple:1147:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(tuple<_Tp...>&& __t) _NOEXCEPT
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/tuple:1157:1: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
get(const tuple<_Tp...>&& __t) _NOEXCEPT
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/tuple:1206:16: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1& get(tuple<_Args...>& __tup) noexcept
               ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/tuple:1213:22: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1 const& get(tuple<_Args...> const& __tup) noexcept
                     ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/tuple:1220:17: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1&& get(tuple<_Args...>&& __tup) noexcept
                ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/tuple:1227:23: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_T1'
constexpr _T1 const&& get(tuple<_Args...> const&& __tup) noexcept
                      ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:1487:59: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
constexpr variant_alternative_t<_Ip, variant<_Types...>>& get(
                                                          ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:1497:60: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
constexpr variant_alternative_t<_Ip, variant<_Types...>>&& get(
                                                           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:1507:65: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
constexpr const variant_alternative_t<_Ip, variant<_Types...>>& get(
                                                                ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:1517:66: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Ip'
constexpr const variant_alternative_t<_Ip, variant<_Types...>>&& get(
                                                                 ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:1527:16: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Tp'
constexpr _Tp& get(variant<_Types...>& __v) {
               ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:1535:17: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Tp'
constexpr _Tp&& get(variant<_Types...>&& __v) {
                ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:1544:22: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Tp'
constexpr const _Tp& get(const variant<_Types...>& __v) {
                     ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:1552:23: note: candidate template ignored: invalid explicitly-specified argument for template parameter '_Tp'
constexpr const _Tp&& get(const variant<_Types...>&& __v) {
```

Reply:
Thanks a lot for your explanation . I will go through it and make neccessary changes !